### PR TITLE
Decipher RecipeProvider-related mappings

### PR DESCRIPTION
**Pull request by 3TUSK**
Did some research on `RecipeProvider` and it was out of my surprise.  
I feel like something is wrong with my current setup; specifically, some local var. aren't correctly renamed. If this PR breaks something, let me know ASAP.  

Other than those, there is still one question left: I am not sure what should interface `gu` and its implementations be called now. It does the following:

  - `gu.a()` -> `gu.toJson`: get a valid recipe json in form of `JsonObject` from GSON. Delegate to `gu.a(JsonObject)` below after writing `Identifier`.
  - `gu.a(JsonObject)` -> `gu.serialize`: write things to that `JsonObject`, see also `BlockEntity.toTag`, but this one returns `void`.
  - `gu.b` -> `gu.getRecipeId`: getter of the recipe id.
  - `gu.c` -> `gu.getSerializer`: getter of corresponding `RecipeSerializer<?>`.
  - `gu.d` -> `gu.toAdvancementJson`: get a valid advancement json in form of `JsonObject` from GSON. This advancement seems to be used for unlocking the recipe defined in `toJson`.
  - `gu.e` -> `gu.getAdvancementId`: getter of id of that advancement.

`gu` is implemented by:

  - `gw$a`, who is in charge of generating shaped recipes.
  - `gx$a`, who is in charge of generating shapeless recipes.
  - `gy$a`, who is in charge of generating smelting/blasting recipes.
  - `gz$a`, who is in charge of generating recipes with just one input.
  - `ha$a`, who is in charge of "generating" all "dynamic" recipes (armor dyeing, map cloning, book duplication, etc.).

For me I would like to call `gu` as `RecipeJsonBuilder`, and thus `ShapedRecipeBuilder$Instance`, `ShapelessRecipeBuilder$Instance`, etc. - however those names sound very strange (esp. `gw`, `gx`, `gy`, `gz` and `ha` are in fact (to a certain extent) builder of builder - `BuilderFactory`?), so I am open for suggestions.

### CODE REVIEW

**Comment by liach** [comment]
Just commit your proposed names instead of asking in pr as if this is an issue.

All names sound great, except `serialize` method is better named `fromJson` instead.

**Review by liach — APPROVED** [review]

**Comment by 3TUSK** [comment]
> except `serialize` method is better named `fromJson` instead.

Actually no, that `serialize` *writes* things to JSON rather than reads things from JSON.

**Comment by liach** [comment]
So `toJson`.

**Comment by 3TUSK** [comment]
Decide to call `gu` as `RecipeJsonProvider` and thus other classes are `XXXFactory` and `XXXFactory$XXXProvider` respectively. This concludes the PR. Comments wanted.

File: mappings/net/minecraft/data/server/recipe/ComplexRecipeJsonFactory.mapping [hunk 1]
```diff
@@ -0,0 +1,15 @@
+CLASS ha net/minecraft/data/server/recipe/ComplexRecipeJsonFactory
+	CLASS ha$1
+		METHOD a (Lcom/google/gson/JsonObject;)V
+			ARG 1 arg1
```
Comment (asiekierka) [inline mappings/net/minecraft/data/server/recipe/ComplexRecipeJsonFactory.mapping:4]: These shouldn't be mapped. I wonder why they got mapped?

File: mappings/net/minecraft/data/server/recipe/CookingRecipeJsonFactory.mapping [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			ARG 1 arg1
+			ARG 2 var2
+			ARG 3 var4
+			ARG 4 var6
+	FIELD a output Laxh;
+	FIELD b input Lazt;
+	FIELD c exp F
+	FIELD d time I
+	FIELD e builder Ln$a;
+	FIELD f gruop Ljava/lang/String;
```
Comment (asiekierka) [inline mappings/net/minecraft/data/server/recipe/CookingRecipeJsonFactory.mapping]: group

File: mappings/net/minecraft/data/server/recipe/ComplexRecipeJsonFactory.mapping (hunk 1 above)
Comment (3TUSK) [reply mappings/net/minecraft/data/server/recipe/ComplexRecipeJsonFactory.mapping:4]: ~~Those are local vars. I heard about the local var. mapping are broken after I mapped those things. How to revert?~~

Edit: wait there aren't any local vars in that method...

Comment (asiekierka) [reply mappings/net/minecraft/data/server/recipe/ComplexRecipeJsonFactory.mapping:4]: Don't worry about it.

**Pull request merged**